File: Medium/113 - PathSumII/Solution.cs
/**
 * Question 113. Path Sum II (https://leetcode-cn.com/problems/path-sum-ii/)
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

/// <summary>
/// 执行用时：316 ms, 在所有 C# 提交中击败了 15.87% 的用户
/// 内存消耗：36.2 MB, 在所有 C# 提交中击败了 7.14% 的用户
/// 
/// This solution is similar to the DFS approach in the [sample solution](https://leetcode-cn.com/problems/path-sum-ii/solution/lu-jing-zong-he-ii-by-leetcode-solution/).
/// </summary>
public class Solution
{
    private List<IList<int>> paths = new List<IList<int>>();

    public IList<IList<int>> PathSum(TreeNode root, int sum)
    {
        this.FindPath(root, sum, new List<int>(), 0);
        return this.paths;
    }

    private void FindPath(TreeNode root, int targetSum, List<int> currentPath, int currentSum)
    {
        if (root == null) { return; }

        int newSum = currentSum + root.val;
        currentPath.Add(root.val);
        // Console.WriteLine($"newSum = {newSum}, targetSum = {targetSum}, this.paths.Count = {this.paths.Count}, root.val = {root.val}");

        if (newSum == targetSum && root.left == null && root.right == null)
        {
            this.paths.Add(currentPath);
            // Console.WriteLine($"newSum = {newSum}, targetSum = {targetSum}, this.paths.Count = {this.paths.Count}, root.val = {root.val}");
            return;
        }

        this.FindPath(root.left, targetSum, currentPath.Select(node => node).ToList(), newSum);
        this.FindPath(root.right, targetSum, currentPath.Select(node => node).ToList(), newSum)
    }
}

File: Medium/113 - PathSumII/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * 执行用时：128 ms, 在所有 typescript 提交中击败了 20.69% 的用户
 * 内存消耗：45.7 MB, 在所有 typescript 提交中击败了 40.00% 的用户
 *
 * Acknowledgement: This solution borrows idea from the BFS approach of the [official solution](https://leetcode-cn.com/problems/path-sum-ii/solution/lu-jing-zong-he-ii-by-leetcode-solution/).
 *
 * @param {(TreeNode | null)} root
 * @param {number} sum
 * @returns {number[][]}
 */
function pathSum(root: TreeNode | null, sum: number): number[][] {
  if (root === null) {
    return [];
  }

  const nodeQueue: TreeNode[] = [root];
  const nodeValueQueue: number[] = [0];

  const hierarchyMap: Map<TreeNode, TreeNode> = new Map<TreeNode, TreeNode>();
  const paths: number[][] = [];

  while (nodeQueue.length) {
    const currentNode: TreeNode = nodeQueue.shift() as TreeNode;
    const currentValue: number =
      (nodeValueQueue.shift() as number) + currentNode.val;

    if (currentNode.left === null && currentNode.right === null) {
      if (currentValue === sum) {
        paths.push(getPath(currentNode, hierarchyMap));
      }
    } else {
      if (currentNode.left) {
        nodeQueue.push(currentNode.left);
        nodeValueQueue.push(currentValue);
        hierarchyMap.set(currentNode.left, currentNode);
      }

      if (currentNode.right) {
        nodeQueue.push(currentNode.right);
        nodeValueQueue.push(currentValue);
        hierarchyMap.set(currentNode.right, currentNode);
      }
    }
  }

  return paths;
}

function getPath(node: TreeNode, hierarchy: Map<TreeNode, TreeNode>): number[] {
  let path: number[] = [];

  while (hierarchy.has(node)) {
    path = [node.val, ...path];
    node = hierarchy.get(node) as TreeNode;
  }

  path = [node.val, ...path];

  // console.log(path);

  return path;
}
